File: tests/test_market.js
import MarketDataService from '../backend/services/MarketDataService.js';

describe('MarketDataService', () => {
  let service;

  beforeAll(() => {
    service = new MarketDataService({ intervalMs: 3600000 });
    service.tick();
  });

  afterAll(() => {
    service.shutdown();
  });

  test('produces analytics with indicators', () => {
    const analytics = service.getAnalytics();
    expect(analytics.assets.length).toBeGreaterThan(0);
    const asset = analytics.assets[0];
    expect(asset).toHaveProperty('rsi');
    expect(asset).toHaveProperty('macd');
    expect(asset).toHaveProperty('bollinger');
    expect(asset.rsi).toBeGreaterThanOrEqual(0);
    expect(asset.macd).toHaveProperty('macd');
  });

  test('maintains rolling history for assets', () => {
    const [first] = service.assets;
    const history = service.getHistory(first.symbol);
    expect(history.length).toBeGreaterThan(0);
    expect(history[0]).toHaveProperty('open');
    expect(history[0]).toHaveProperty('close');
  });
});

File: tests/test_market.py
"""Unit tests for the market data generator."""

from __future__ import annotations

from pybackend.services.market_data import MarketDataService


def test_produces_analytics_with_indicators():
    service = MarketDataService(interval_ms=3_600_000)
    service.tick()
    analytics = service.get_analytics()
    assert len(analytics["assets"]) > 0
    asset = analytics["assets"][0]
    assert asset["rsi"] is None or asset["rsi"] >= 0
    assert asset["macd"] is None or "macd" in asset["macd"]


def test_maintains_history():
    service = MarketDataService(interval_ms=3_600_000)
    service.tick()
    symbol = service.assets[0]["symbol"]
    history = service.get_history(symbol)
    assert len(history) > 0
    candle = history[0]
    assert {"open", "close", "high", "low"}.issubset(candle.keys())
